phpBB/phpbb/captcha/plugins/turnstile.php

```diff
@@ -0,0 +1,288 @@
+use phpbb\config\config;
+use phpbb\db\driver\driver_interface;
+use phpbb\language\language;
+use phpbb\log\log_interface;
+use phpbb\request\request_interface;
+use phpbb\template\template;
+use phpbb\user;
+
+class turnstile extends base
+{
+	/** @var string URL to cloudflare turnstile API javascript */
+	private const SCRIPT_URL = 'https://challenges.cloudflare.com/turnstile/v0/api.js';
+
+	/** @var string API endpoint for turnstile verification */
+	private const VERIFY_ENDPOINT = 'https://challenges.cloudflare.com/turnstile/v0/siteverify';
+
+	/** @var Client */
+	protected Client $client;
+
+	/** @var language */
+	protected language $language;
+
+	/** @var log_interface */
+	protected log_interface $log;
+
+	/** @var template */
+	protected template $template;
+
+	/** @var string Service name */
+	protected string $service_name = '';
+
+	/** @var array|string[] Supported themes for Turnstile CAPTCHA */
+	protected static array $supported_themes = [
+		'light',
+		'dark',
+		'auto'
+	];
+
+	/**
+	 * Constructor for turnstile captcha plugin
+	 *
+	 * @param config $config
+	 * @param driver_interface $db
+	 * @param language $language
+	 * @param log_interface $log
+	 * @param request_interface $request
+	 * @param template $template
+	 * @param user $user
+	 */
+	public function __construct(config $config, driver_interface $db, language $language, log_interface $log, request_interface $request, template $template, user $user)
+	{
+		parent::__construct($config, $db, $language, $request, $user);
+
+		$this->language = $language;
+		$this->log = $log;
+		$this->template = $template;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function is_available(): bool
+	{
+		$this->init($this->type);
+
+		return !empty($this->config->offsetGet('captcha_turnstile_sitekey'))
+			&& !empty($this->config->offsetGet('captcha_turnstile_secret'));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function has_config(): bool
+	{
+		return true;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function get_name(): string
+	{
+		return 'CAPTCHA_TURNSTILE';
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function set_name(string $name): void
+	{
+		$this->service_name = $name;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function init(confirm_type $type): void
+	{
+		parent::init($type);
+
+		$this->language->add_lang('captcha_turnstile');
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function validate(): bool
+	{
+		if (parent::validate())
+		{
+			return true;
+		}
+
+		$turnstile_response = $this->request->variable('cf-turnstile-response', '');
+		if (!$turnstile_response)
+		{
+			// Return without checking against server without a turnstile response
+			return false;
+		}
+
+		// Retrieve form data for verification
+		$form_data = [
+			'secret'			=> $this->config['captcha_turnstile_secret'],
+			'response'			=> $turnstile_response,
+			'remoteip'			=> $this->request->header('CF-Connecting-IP'),
+			//'idempotency_key'	=> $this->confirm_id, // check if we need this
```

Comment:
Remaining todo, please remove if not necessary.

Reply:
i don't think we want to allow multiple validations.

Reply:
Yeah, I agree that this shouldn't be needed and will remove it.